modules/xml/tests/test_xml_parser.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2017 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ */
+
+
+#include <criterion/criterion.h>
+#include <criterion/parameterized.h>
+#include "msg_parse_lib.h"
+#include "xml.h"
+#include "apphook.h"
+#include "scratch-buffers.h"
+
+void
+setup(void)
+{
+  configuration = cfg_new_snippet(0x0390);
+  app_startup();
+}
+
+void
+teardown(void)
+{
+  scratch_buffers_explicit_gc();
+  app_shutdown();
+  cfg_free(configuration);
+}
+
+TestSuite(xmlparser, .init = setup, .fini = teardown);
+
+typedef struct
+{
+  const gchar *input;
+  const gchar *key;
+  const gchar *value;
+} ValidXMLTestCase;
+
+ParameterizedTestParameters(xmlparser, valid_inputs)
+{
+  static ValidXMLTestCase test_cases[] =
+  {
+    {"<tag1>value1</tag1>", ".xml.tag1", "value1"},
+    {"<tag1 attr='attr_value'>value1</tag1>", ".xml.tag1._attr", "attr_value"},
+    {"<tag1><tag2>value2</tag2></tag1>", ".xml.tag1.tag2", "value2"},
+    {"<tag1>part1<tag2>value2</tag2>part2</tag1>", ".xml.tag1", "part1part2"},
+    {"<tag1><tag11></tag11><tag12><tag121>value</tag121></tag12></tag1>", ".xml.tag1.tag12.tag121", "value"},
+    {"<tag1><tag11></tag11><tag12><tag121 attr1='1' attr2='2'>value</tag121></tag12></tag1>", ".xml.tag1.tag12.tag121._attr1", "1"},
+    {"<tag1><tag11></tag11><tag12><tag121 attr1='1' attr2='2'>value</tag121></tag12></tag1>", ".xml.tag1.tag12.tag121._attr2", "2"},
```

Comment:
I was thinking about the following case:

```xml
<tag>
   <tag1>value11</tag1>
   <tag2>value22</tag1>
</tag>
```

What should be the result of `tag.tag1` ?

Reply:
There is a drop_invalid() with default value: no. User can control this behaviour.

For the first part: if we write tag1 instead of tag2 as well:
```
<tag>
   <tag1>value11</tag1>
   <tag1>value22</tag1>
</tag>
```
the result will be concatenated. The same behaviour is visible with xmllint:
$ xmllint --xpath "/tag/tag1/text()" text.xml
value11value22%
$
It is currently a limitation of the parser users cannot access the individual tags in such vector-like structures. This limitation will be documented in the administrator's guide.